docs/contributing/api-mocking.md

```diff
@@ -1,53 +1,176 @@
 # API Mocking
 
-Testing functions that interact with the API can be complicated because the API can have many different responses or errors, but the behavior is defined outside this app.  The code that calls the API needs to be tested in a predictable and reproducible way.
+API testing can be complicated because the API can have many different responses
+or errors. Additionally the API behavior is generally outside the scope of the
+test.  APIs can have errors, they may be stateful, there may be network
+instability and a host of other issues that conspire to cause test failures.
+Unit tests shouldn't make any API calls because it makes the test slow, brittle
+and non-determistic which undermines many of the advantages of unit tests such
+as using them for CI on every change.
+
+The code that calls the API still needs to be tested in a predictable and
+reproducible way.  We need a way to remove all that uncertainty without
+introducing burdensome complexity in our testing environment such as running an
+entire application and networking stack.
+
+The solution is mock the API.  This means that we will tell the test harness how
+to interpret calls to the API so that the code under test can experience
+different scenarios.  This is not a fake API, which would seek to mimic the API
+behavior, but rather a collection of expected requests and responses.
+
+## Usage
+
+Mock API usage is a matter of setting expectations of calls that will be made
+and stating what those calls should return.  This tells the test how to handle
+the expected calls and that anything else is unexpected.  This is important to
+guard against accidental calls so you never have to say, "I didn't know that
+function had stateful behavior!"
+
+Expectations are added through helpers, such as `returnsOnce`.  Each of these
+take in a request to match against (from the effect) and then a behavior to
+execute when the request is made.
+
+An expectation is considered _satisfied_ when it has been used to fulful the
+minimum number it expects.  An expectation is considered _exhausted_ if has
+fulfulled the maximum number of requests it expects.
```

Comment:
Add two points here to prime the user for the details later:
- At the end of a test, all expectations must be satisfied (no expectation may go unused)
- Trying to reuse an exhausted expectation will raise an error

Reply:
Maybe a third point:
- expectation do NOT have to be exhausted at the end of a test